modules/packages/ZMQ.chpl

```diff
@@ -728,6 +735,93 @@ module ZMQ {
       }
     }
 
+    /*
+      Get socket options;
+      see `zmq_getsockopt <http://api.zeromq.org/4-0:zmq-getsockopt>`_
+
+      :arg option: a socket option;
+          e.g., :const:`LINGER`, :const:`LAST_ENDPOINT`
+      :type option: `int`
+
+      :returns: Varies depending on the option specified, see the link above.
+
+      :throws IllegalArgumentError: Thrown when the option provided is not
+                                    supported for getsockopt
+     */
+    proc getsockopt(param option: int) throws where (getsockoptRetTypeHelper(option)
```

Comment:
If we stick with the current approach, I'd strongly suggest implementing these overloads using a form like:

```proc getsockopt(param option: int): string throws where (option == LAST_ENDPOINT) {```

and then permitting all the (legal) overloads to be chpldoc'd because (a) it seems more straightforward and (b) it doesn't put a reference to a nodoc'd routine into  the chpldocumentation for this routine.  That is, it'll make the chpldoc output easier to read (and more self-sufficient) by permitting me to see that if I pass in this argument, I'll get this type; and if I pass in this other one, I'll get this other type.

Reply:
> it doesn't put a reference to a nodoc'd routine into the chpldocumentation for this routine

Are you referring to the where clause?  #5950 shows that we don't support those in chpldoc just yet.

Does there being 70-ish options for this routine change your thinking at all?  My figuring was that the ZMQ getsockopt API page referenced would be mostly sufficient for covering which types are returned when, but I can also do an individual version.

Reply:
> Are you referring to the where clause? #5950 shows that we don't support those in chpldoc just yet.

Yeah, good point.  But we want them to be, right?  (i.e., someday this will be?).

> Does there being 70-ish options for this routine change your thinking at all?

It gives me pause, but just a little...  If it wasn't clear, I wouldn't have a separate overload for each option that returns a string, but would simply make the where clause list all the cases that matched that return type:

```chapel
proc getsockopt(param option: int): string throws where (option == LAST_ENDPOINT ||
                                                          option == PLAIN_USERNAME) { ... }
```

If this got too unwieldy, a middle-of-the-road option would be to have a helper function per return type:

```chapel
proc getsockopt(param option: int): string throws where optionReturnsString(option) { ... }
```

which is at least a bit more self-documenting...  But I probably wouldn't go down this path until the where clauses became obviously unwieldy.

Reply:
I think I'd prefer making getsockoptRetTypeHelper public (potentially with a different name) to exposing a series of optionReturnsX functions.  That way the user can just call one function to determine the expected return type for a given option, instead of trial and error until they get a `true`.

Reply:
> But we want them to be, right?

Yup!  Though I am starting to wonder if it would be good to have ways to turn information on and off in the documentation (for instance, we document the value of global variables if they are provided in the declaration, but I didn't necessarily want to for `LINGER`.  It seems like things would get too finicky if you have to say "display the type but not the default value, or the where clause for this function but not for this other one".  But I'm probably overthinking it)

Reply:
> I think I'd prefer making getsockoptRetTypeHelper public

If we went that route, I'm not sure why we wouldn't just use `: sockoptToType(option)` as the routine's return type rather than making it generic and putting a where clause on the routine.

But I think I still like the `where option == ...` approach best for simplicity and straightforwardness.  How many options return each possible return type?  If I was scanning correctly, it looked like maybe only 3 returned strings?  If it was the case that 95% returned int (say) and only a few returned other types, a where-clause-free version could be used for that 95% case.

Reply:
Or... maybe the `: sockoptToType(option)` approach is best if the differences between the routines for different return types were minimal and it allowed us to write the whole thing as one routine rather than one per return type for easier maintainability?

Reply:
It looks like 7 return character strings, though there are a few additional ones that return "binary data or Z85 text string".  I think most of the rest are integers, but some are integers that are really booleans.

Hmm, I think the `: sockoptToType(option)` is best.  The difference should be pretty minimal, just which helper C function gets called.